Jason/scheduler

jasonjewik opened the pull request:
### Given a list of task objects as defined by the the task model, set their `scheduled_time` to be within the user's hours of operation and before the latest task's due date.

_Important Notes_:
* The algorithm leaves the `scheduled_time` field as `null`, if no suitable time could be found. (For example, the task takes longer than any given free interval, or the task cannot be scheduled before its due date.
* This PR does NOT include code to actually create Google Calendar events, nor code to permanently change the task objects. The POST request returns a copy of the task list that is modified with the suggested scheduled times.

**Request looks like:**
```
URL: localhost:31337/api/schedule
Body: {
	"ids": [{{task_ids}}],
	"timeMin": "{{timeMin}}",
	"timeZone": "America/Los_Angeles"
}
```

**Returns:**
```
{
  "data": [
    {
      "id": 7,
      "user_id": "1",
      "name": "input",
      "description": "If we input the interface, we can get to the SDD driver through the bluetooth IB driver!",
      "duration": 5879,
      "due_date": "2020-02-12T23:00:00.000Z",
      "completed": false,
      "completed_time": null,
      "scheduled": false,
      "scheduled_time": null,
      "calendar_id": null,
      "event_id": null,
      "start_time": null,
      "end_time": null,
      "created_time": "2020-02-12T23:32:51.311Z",
      "updated_time": "2020-02-12T23:32:51.311Z"
    },
    {
      "id": 1,
      "user_id": "1",
      "name": "navigate",
      "description": "Try to index the JSON interface, maybe it will navigate the back-end capacitor!",
      "duration": 2005,
      "due_date": "2020-02-13T23:31:12.611Z",
      "completed": false,
      "completed_time": null,
      "scheduled": false,
      "scheduled_time": "2020-02-13T07:00:00-08:00",
      "calendar_id": null,
      "event_id": null,
      "start_time": null,
      "end_time": null,
      "created_time": "2020-02-12T23:31:12.737Z",
      "updated_time": "2020-02-12T23:31:12.737Z"
    },
    {
      "id": 9,
      "user_id": "1",
      "name": "copy",
      "description": "We need to override the online CSS system!",
      "duration": 2348,
      "due_date": "2020-02-20T23:00:00.000Z",
      "completed": false,
      "completed_time": null,
      "scheduled": false,
      "scheduled_time": "2020-02-13T07:33:25-08:00",
      "calendar_id": null,
      "event_id": null,
      "start_time": null,
      "end_time": null,
      "created_time": "2020-02-12T23:32:58.444Z",
      "updated_time": "2020-02-12T23:32:58.444Z"
    },
    {
      "id": 4,
      "user_id": "1",
      "name": "reboot",
      "description": "Try to compress the EXE system, maybe it will compress the cross-platform pixel!",
      "duration": 5849,
      "due_date": "2020-02-21T23:31:19.179Z",
      "completed": false,
      "completed_time": null,
      "scheduled": false,
      "scheduled_time": "2020-02-13T08:12:33-08:00",
      "calendar_id": null,
      "event_id": null,
      "start_time": null,
      "end_time": null,
      "created_time": "2020-02-12T23:31:19.267Z",
      "updated_time": "2020-02-12T23:31:19.267Z"
    },
    {
      "id": 3,
      "user_id": "1",
      "name": "reboot",
      "description": "If we copy the sensor, we can get to the COM application through the solid state USB capacitor!",
      "duration": 2607,
      "due_date": "2020-02-22T23:31:16.857Z",
      "completed": false,
      "completed_time": null,
      "scheduled": false,
      "scheduled_time": "2020-02-13T09:50:02-08:00",
      "calendar_id": null,
      "event_id": null,
      "start_time": null,
      "end_time": null,
      "created_time": "2020-02-12T23:31:16.955Z",
      "updated_time": "2020-02-12T23:31:16.955Z"
    },
    {
      "id": 10,
      "user_id": "1",
      "name": "input",
      "description": "If we reboot the pixel, we can get to the CSS transmitter through the back-end JBOD bus!",
      "duration": 4936,
      "due_date": "2020-02-24T23:00:00.000Z",
      "completed": false,
      "completed_time": null,
      "scheduled": false,
      "scheduled_time": "2020-02-13T10:33:29-08:00",
      "calendar_id": null,
      "event_id": null,
      "start_time": null,
      "end_time": null,
      "created_time": "2020-02-12T23:32:59.526Z",
      "updated_time": "2020-02-12T23:32:59.526Z"
    },
    {
      "id": 2,
      "user_id": "1",
      "name": "navigate",
      "description": "generating the sensor won't do anything, we need to parse the neural SAS bus!",
      "duration": 6217,
      "due_date": "2020-02-24T23:31:15.199Z",
      "completed": false,
      "completed_time": null,
      "scheduled": false,
      "scheduled_time": "2020-02-14T17:00:00-08:00",
      "calendar_id": null,
      "event_id": null,
      "start_time": null,
      "end_time": null,
      "created_time": "2020-02-12T23:31:15.294Z",
      "updated_time": "2020-02-12T23:31:15.294Z"
    },
    {
      "id": 5,
      "user_id": "1",
      "name": "quantify",
      "description": "You can't hack the bus without connecting the back-end COM array!",
      "duration": 1910,
      "due_date": "2020-02-24T23:31:21.387Z",
      "completed": false,
      "completed_time": null,
      "scheduled": false,
      "scheduled_time": "2020-02-13T14:00:00-08:00",
      "calendar_id": null,
      "event_id": null,
      "start_time": null,
      "end_time": null,
      "created_time": "2020-02-12T23:31:21.480Z",
      "updated_time": "2020-02-12T23:31:21.480Z"
    },
    {
      "id": 6,
      "user_id": "1",
      "name": "reboot",
      "description": "The TCP protocol is down, override the neural array so we can copy the AI system!",
      "duration": 2976,
      "due_date": "2020-02-25T23:31:24.116Z",
      "completed": false,
      "completed_time": null,
      "scheduled": false,
      "scheduled_time": "2020-02-13T16:15:00-08:00",
      "calendar_id": null,
      "event_id": null,
      "start_time": null,
      "end_time": null,
      "created_time": "2020-02-12T23:31:24.201Z",
      "updated_time": "2020-02-12T23:31:24.201Z"
    },
    {
      "id": 8,
      "user_id": "1",
      "name": "parse",
      "description": "Try to quantify the ADP system, maybe it will reboot the auxiliary hard drive!",
      "duration": 3260,
      "due_date": "2020-02-26T23:00:00.000Z",
      "completed": false,
      "completed_time": null,
      "scheduled": false,
      "scheduled_time": "2020-02-14T07:00:00-08:00",
      "calendar_id": null,
      "event_id": null,
      "start_time": null,
      "end_time": null,
      "created_time": "2020-02-12T23:32:56.239Z",
      "updated_time": "2020-02-12T23:32:56.239Z"
    }
  ],
  "error": null
}

```

**Postman Screenshot**
![image](https://user-images.githubusercontent.com/13477221/74408511-7abac280-4de9-11ea-985e-187eae91979c.png)

jasonjewik commented:
> TODO: Fix algorithm because right now it skips days sometimes.

Fixed. Other TODO's are noted as comments in the code itself.

alexlongerbeam commented on src/controllers/schedule.js:
```diff
@@ -0,0 +1,281 @@
+                moment(interval[0]).tz(req.body.timeZone), 
+                moment(interval[1]).tz(req.body.timeZone)
+            ];
+        });
```
This is some clean code right here, love it

alexlongerbeam commented on src/controllers/schedule.js:
```diff
@@ -0,0 +1,281 @@
+        // Create free intervals from busy intervals in user's local time
+        let freeInts = [];
+        let lastEndTime = moment(req.body.timeMin).tz(req.body.timeZone);
+        busyMomInts.forEach((busyInt) => {
```
This should go into `ScheduleService.js` -- the task here is convert busy times to free times, using user's hours of operation. You can make busy times and hours of operation all parameters to the function so `ScheduleService.js` does not have to fetch the user model again

alexlongerbeam commented on src/controllers/schedule.js:
```diff
@@ -0,0 +1,281 @@
+        //      freeTime - An array of two-element arrays, each representing an interval of free time.
+        //          The times should be in seconds since epoch.
+        function solve(tasks, freeTime) {
+            if (tasks.length == 0)
```
Additionally, this whole flow (schedule tasks given free times) should be its own function in `ScheduleService.js`

alexlongerbeam commented on src/controllers/schedule.js:
```diff
@@ -0,0 +1,281 @@
+            if (dt != null) {
+                dt = moment.unix(dt);
+                scheduledTaskList[i].scheduled_time = dt.tz(req.body.timeZone).format();
+            }
```
I think this is fine to stay here

alexlongerbeam commented on src/controllers/schedule.js:
```diff
@@ -0,0 +1,281 @@
+const express = require('express');
+const moment = require('moment-timezone');
+
+const ScheduleController = (userModel, taskModel, authService, googleAPIService, solve) => {
```
Not sure this controller needs `solve` as a parameter

jasonjewik replied:
Oh, haha. I didn't realize that was still there! I actually originally had `solve` in a separate file, which I undid before the commit.

alexlongerbeam commented on src/models/task/postgres.js:
```diff
@@ -8,15 +8,15 @@ const TaskRepo = (postgres) => {
+            start_time timestamptz,
+            end_time timestamptz,
+            created_time timestamptz DEFAULT NOW(),
+            updated_time timestamptz DEFAULT NOW()
```
Not saying this is wrong, but what is the reason behind all this? Do we need to store timezone info if our contract is that they will be stored in UTC? Or is this needed for Moment?

jasonjewik replied:
Currently, all my Moment math is done in the user's local time zone because that made it easier for me to manually compare the POST request result to my actual calendar and debug the algorithm. I'll convert everything to be done in UTC before I merge into `master`.

jasonjewik commented:
Updated code to reflect the requested changes. Also squashed a bug that caused events to be scheduled at the same time.

New sample output looks like:
```
{
    "data": [
        {
            "id": 2,
            "user_id": "1",
            "name": "connect",
            "description": "The JBOD application is down, back up the cross-platform array so we can copy the PCI driver!",
            "duration": 90,
            "due_date": "2020-02-19T07:00:00.000Z",
            "completed": false,
            "completed_time": null,
            "scheduled": false,
            "scheduled_time": "2020-02-18T09:00:00-08:00",
            "calendar_id": null,
            "event_id": null,
            "start_time": null,
            "end_time": null,
            "created_time": "2020-02-18T07:38:02.363Z",
            "updated_time": "2020-02-18T07:38:02.363Z"
        },
        {
            "id": 7,
            "user_id": "1",
            "name": "back up",
            "description": "I'll override the haptic JBOD card, that should hard drive the SDD matrix!",
            "duration": 102,
            "due_date": "2020-02-20T07:00:00.000Z",
            "completed": false,
            "completed_time": null,
            "scheduled": false,
            "scheduled_time": "2020-02-18T14:00:00-08:00",
            "calendar_id": null,
            "event_id": null,
            "start_time": null,
            "end_time": null,
            "created_time": "2020-02-18T07:38:07.148Z",
            "updated_time": "2020-02-18T07:38:07.148Z"
        },
        {
            "id": 10,
            "user_id": "1",
            "name": "back up",
            "description": "Try to index the HDD feed, maybe it will override the 1080p program!",
            "duration": 62,
            "due_date": "2020-02-21T07:00:00.000Z",
            "completed": false,
            "completed_time": null,
            "scheduled": false,
            "scheduled_time": "2020-02-18T10:30:00-08:00",
            "calendar_id": null,
            "event_id": null,
            "start_time": null,
            "end_time": null,
            "created_time": "2020-02-18T07:38:10.141Z",
            "updated_time": "2020-02-18T07:38:10.141Z"
        },
        {
            "id": 6,
            "user_id": "1",
            "name": "parse",
            "description": "connecting the card won't do anything, we need to navigate the cross-platform COM driver!",
            "duration": 36,
            "due_date": "2020-02-22T07:00:00.000Z",
            "completed": false,
            "completed_time": null,
            "scheduled": false,
            "scheduled_time": "2020-02-18T07:00:00-08:00",
            "calendar_id": null,
            "event_id": null,
            "start_time": null,
            "end_time": null,
            "created_time": "2020-02-18T07:38:06.296Z",
            "updated_time": "2020-02-18T07:38:06.296Z"
        },
        {
            "id": 3,
            "user_id": "1",
            "name": "override",
            "description": "I'll back up the redundant RSS panel, that should matrix the RSS program!",
            "duration": 76,
            "due_date": "2020-02-23T07:00:00.000Z",
            "completed": false,
            "completed_time": null,
            "scheduled": false,
            "scheduled_time": "2020-02-18T15:42:00-08:00",
            "calendar_id": null,
            "event_id": null,
            "start_time": null,
            "end_time": null,
            "created_time": "2020-02-18T07:38:03.313Z",
            "updated_time": "2020-02-18T07:38:03.313Z"
        },
        {
            "id": 9,
            "user_id": "1",
            "name": "program",
            "description": "We need to back up the wireless AGP interface!",
            "duration": 64,
            "due_date": "2020-02-24T07:00:00.000Z",
            "completed": false,
            "completed_time": null,
            "scheduled": false,
            "scheduled_time": "2020-02-19T12:00:00-08:00",
            "calendar_id": null,
            "event_id": null,
            "start_time": null,
            "end_time": null,
            "created_time": "2020-02-18T07:38:09.048Z",
            "updated_time": "2020-02-18T07:38:09.048Z"
        },
        {
            "id": 5,
            "user_id": "1",
            "name": "override",
            "description": "We need to transmit the solid state GB interface!",
            "duration": 85,
            "due_date": "2020-02-25T07:00:00.000Z",
            "completed": false,
            "completed_time": null,
            "scheduled": false,
            "scheduled_time": "2020-02-19T16:45:00-08:00",
            "calendar_id": null,
            "event_id": null,
            "start_time": null,
            "end_time": null,
            "created_time": "2020-02-18T07:38:05.301Z",
            "updated_time": "2020-02-18T07:38:05.301Z"
        },
        {
            "id": 1,
            "user_id": "1",
            "name": "index",
            "description": "I'll parse the 1080p IB capacitor, that should port the COM driver!",
            "duration": 92,
            "due_date": "2020-02-26T07:00:00.000Z",
            "completed": false,
            "completed_time": null,
            "scheduled": false,
            "scheduled_time": "2020-02-19T18:10:00-08:00",
            "calendar_id": null,
            "event_id": null,
            "start_time": null,
            "end_time": null,
            "created_time": "2020-02-18T07:38:01.327Z",
            "updated_time": "2020-02-18T07:38:01.327Z"
        },
        {
            "id": 4,
            "user_id": "1",
            "name": "synthesize",
            "description": "We need to synthesize the primary HDD bandwidth!",
            "duration": 98,
            "due_date": "2020-02-28T07:00:00.000Z",
            "completed": false,
            "completed_time": null,
            "scheduled": false,
            "scheduled_time": "2020-02-20T07:00:00-08:00",
            "calendar_id": null,
            "event_id": null,
            "start_time": null,
            "end_time": null,
            "created_time": "2020-02-18T07:38:04.304Z",
            "updated_time": "2020-02-18T07:38:04.304Z"
        },
        {
            "id": 8,
            "user_id": "1",
            "name": "generate",
            "description": "The PCI capacitor is down, synthesize the digital circuit so we can copy the SSL interface!",
            "duration": 54,
            "due_date": "2020-02-28T07:00:00.000Z",
            "completed": false,
            "completed_time": null,
            "scheduled": false,
            "scheduled_time": "2020-02-18T16:58:00-08:00",
            "calendar_id": null,
            "event_id": null,
            "start_time": null,
            "end_time": null,
            "created_time": "2020-02-18T07:38:08.053Z",
            "updated_time": "2020-02-18T07:38:08.053Z"
        }
    ],
    "error": null
}
```

jasonjewik commented on src/services/schedule.js:
```diff
@@ -0,0 +1,198 @@
+                });
+
+                if (newFreeTime[0][0] == currTask.scheduled_time)
+                    newFreeTime.splice(i, 1);
```
This is where the bug in the scheduling algorithm was. Previously, I was removing the first element of `freeTime`, when it should have been the currently examined element.

jasonjewik commented:
**Scheduling:**
![image](https://user-images.githubusercontent.com/13477221/75805079-f7302980-5d35-11ea-8070-df4dd46652d4.png)
Changed the `scheduled_time` field to be an ISO string (UTC timezone).

**Confirming Schedule:**
![image](https://user-images.githubusercontent.com/13477221/75805605-fe0b6c00-5d36-11ea-90b9-36a02a38f485.png)
Sets `start time` and `end_time` to ISO stings. Sets `scheduled` to `true`. Sets `event_id` to be the Google Calendar event ID. 
TODO: Force scheduling.

**Creates Calendar Events:**
![image](https://user-images.githubusercontent.com/13477221/75805085-fa2b1a00-5d35-11ea-8625-c04a092ec12b.png)
